drivers/builtin/src/chacha20.c

```diff
@@ -22,13 +22,121 @@
 
 #include "mbedtls/platform.h"
 
-#define ROTL32(value, amount) \
-    ((uint32_t) ((value) << (amount)) | ((value) >> (32 - (amount))))
-
 #define CHACHA20_CTR_INDEX (12U)
 
 #define CHACHA20_BLOCK_SIZE_BYTES (4U * 16U)
 
+#if defined(MBEDTLS_HAVE_NEON_INTRINSICS)
+
+// Tested on all combinations of armv7 arm/thumb2; armv8 arm/thumb2/aarch64 on clang 14, gcc 11,
+// and some more recent versions.
+
+// Define rotate-left operations that rotate within each 32-bit element in a 128-bit vector.
+static inline uint32x4_t chacha20_neon_vrotlq_16_u32(uint32x4_t v)
+{
+    return vreinterpretq_u32_u16(vrev32q_u16(vreinterpretq_u16_u32(v)));
+}
+
+static inline uint32x4_t chacha20_neon_vrotlq_12_u32(uint32x4_t v)
+{
+    uint32x4_t x = vshlq_n_u32(v, 12);
+    return vsriq_n_u32(x, v, 20);
+}
+
+static inline uint32x4_t chacha20_neon_vrotlq_8_u32(uint32x4_t v)
+{
+    uint32x4_t result;
+#if defined(MBEDTLS_ARCH_IS_ARM64)
```

Comment:
I am a bit concerned on the [DIT bit](https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/159/commits/bd5c34ccf89b6c2033d9b31d5d1b52b51013a35d#r2303621644) requirement for this to be constant time. Do you think we could go for the `else` branch for now and add this part back when we managed to sort out our thoughts around DIT?

Reply:
Note that the look-up table that describes how to re-order the data is a compile-time const - only the register it is shuffling is unknown. So I don't think this could result in data-dependent timing.